### PR TITLE
Environment constructor and `GetTemplate`

### LINKED ISSUE #7: Implement simplified Environment class
Needed methods:
- constructor
- get_template

### PR DESCRIPTION
**Pull request by badochov**
closes #7 
Moved `loaders` to `environement` package to avoid cyclic dep

### CODE REVIEW

File: src/environment/environment.go [hunk 1]
```diff
@@ -1,8 +1,212 @@
...
+	if cacheSize > 0 {
+		return lru.New(cacheSize)
+	}
+	if cacheSize < 0 {
+		return make(mapCache), nil
+	}
+	return nil, nil
+}
+
+type Extension struct{} // TODO change to real extension
```
Comment (jmolinski) [inline src/environment/environment.go:138]: #15 also adds an interface (and a separate package) for extensions

File: src/environment/environment.go [hunk 2]
```diff
@@ -1,8 +1,212 @@
...
+
+type EnvOpts struct {
+	*lexer.EnvLexerInformation
+	Optimized  bool
+	Extensions map[string]func(*Environment) Extension // TODO jinja accepts also extensions names but it's python import magic I don't know how to do it in golang.
```
Comment (jmolinski) [inline src/environment/environment.go:143]: I don't think we need to port that

File: src/utils/maps/maps.go [hunk 3]
```diff
@@ -13,12 +13,39 @@ func Keys[K comparable, V any](m map[K]V) []K {
...
 	keys := Keys(m)
 	sort.Slice(keys, func(i int, j int) bool { return keys[i] < keys[j] })
 	return keys
 }
 
+func Chain[K comparable, V any](ms ...map[K]V) map[K]V {
+	res := make(map[K]V)
+	for _, m := range ms {
+		for k, v := range m {
+			if _, ok := res[k]; !ok {
```
Comment (jmolinski) [inline src/utils/maps/maps.go:34]: shouldn't the last occurrence be preferred?

File: src/environment/environment.go [hunk 4]
```diff
@@ -1,8 +1,212 @@
...
+	Shared        bool
+	Concat        func([]string) string
+	ContextClass  runtime.ContextClass
 	TemplateClass Class
 	*lexer.EnvLexerInformation
+	Optimized  bool
+	Extensions map[string]Extension // extension name or extension
+	Undefined  UndefinedConstructor
+	Finalize   func(...any) any
+	AutoEscape any // bool or func(string)bool
```
Comment (jmolinski) [inline src/environment/environment.go]: Shouldn't we make it a generic with limited possible underlying types?

Or, in the constructor, if we receive a `bool` we could store it as `func(string) bool { return theBool }`

File: src/environment/environment.go (hunk 1 above)
Comment (badochov) [reply src/environment/environment.go:138]: yeah i just added this one as a placeholder

File: src/environment/environment.go (hunk 2 above)
Comment (badochov) [reply src/environment/environment.go:143]: I'll leave this comment for now tho

File: src/utils/maps/maps.go (hunk 3 above)
Comment (badochov) [reply src/utils/maps/maps.go:34]: ChainMap in python prefers the first one

File: src/environment/environment.go (hunk 4 above)
Comment (badochov) [reply src/environment/environment.go]: I thougt about generic but `LinkedTo` makes it impossible, I just wish go would have Union types.

I like your idea about change in constructor though

**Review by jmolinski — APPROVED** [review]
LGTM, great

**Pull request merged**